synapse/handlers/federation.py

```diff
@@ -2009,6 +2013,17 @@ async def on_send_membership_event(
         # the room, so we send it on their behalf.
         event.internal_metadata.send_on_behalf_of = origin
 
+        # Sign the event since we're vouching on behalf of the remote server that
+        # the event is valid to be sent into the room.
+        event.signatures.update(
+            compute_event_signature(
+                room_version,
+                event.get_pdu_json(),
+                self.hs.hostname,
+                self.hs.signing_key,
+            )
+        )
```

Comment:
We discussed this in #synapse-dev:matrix.org and the decision was to unconditionally do this (for `send_{join,leave,knock}`) since it sounds like a good idea. I don't know if we should only do it for the new room version though? The MSC doesn't assert it must always be sent, just that it needs to be done for join events in restricted rooms.

Reply:
I'm a little bit nervous about adding signatures that we then don't check? The problem being if in future these signatures can no longer be verified, what do servers do?

Reply:
They would do the same thing as if the current signature is not able to be verified, I think?

Reply:
Synapse doesn't check extraneous signatures, so in some ways that's fine, but its a bit odd for Synapse to add extra signatures that aren't needed and I do think will likely cause problems where new servers think they should be verifying those sigs :shrug: